content/concepts/address-discovery-and-advertisement.md

```diff
@@ -0,0 +1,216 @@
+All the methods and address sets discussed above go into making the final set of a node’s dialable addresses.
+The way a libp2p peer shares these dialable addresses with other peers is:
+
+- When a peer connects to other peers, it asks it’s `Host` for it’s current set of dialable addresses.
+
+- This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).
+
+- Put those addresses in the `listenAddrs` field of the `Identify` message and send the `Identify` message to
+  the other peer.
+
+- When a peer receives another peer’s dialable addresses from the `listenAddrs` field of the `Identify` message,
+  it stores them in it’s [peerstore][peerstore_repo] for some time. The timing does NOT matter for this discussion.
+
+- Any peer can then lookup a peer’s addresses by asking the DHT network to find the peer with that [peerID][peer_id] by
+  using the [FindPeer][find_peer] DHT API. The DHT nodes will use the addresses stored in their peerstore as described above
+  to find and return the addresses of the peer.
+
+    #### Handling changes to the peer’s dialable addresses
+
+    All well and good so far. But the set of a peer’s dialable addresses is ofcourse not a static set. A peer can
+    start listening on new interfaces, see more of it’s observed addresses, change it’s network, be assigned a new
+    IP address by the Router, etc. Hence, we also need a mechanism to detect changes to a peer’s dialable addresses set
+    and advertise these changes to the world on the fly. The way we do this is by
+    **running an event loop** in the `Host`.
+
+
+    #### When does the Host event loop fire ?
+    - Every 5 seconds.
+
+    - When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.
+
+    - When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a
+    connection with a new relay server or loses connection with an existing relay server since all of these would
+    cause a change in it’s relay addresses.
+
+    #### What does the Host do when the event loop fires ?
+    - Updates it’s primary network interface address by using the netroute library as discussed above
+    (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.
+
+    - Builds a new address set by using all the addresses discovered by the various mechanisms described in the
+    `Address Discovery` section and compares it with the previous address set.
+
+    - If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
+
+    - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all
+    peers it is connected to via the `Identify Push` protocol.
+    The ID service is the same service that also handles the `Identify` protocol described above.
```

Comment:
```suggestion
- On interval, every 5 seconds.

- When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.

- When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a connection with a new relay server, or loses connection with an existing relay server since all of these would cause a change in its relay addresses.

##### What does the Host do when the event loop fires ?
- Updates its primary network interface address by using the netroute library as discussed above (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.

- Builds a new address set by using all the addresses discovered by the various mechanisms described in the `Address Discovery` section and compares it with the previous address set.

- If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
    - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all peers it is connected to via the `Identify Push` protocol. The ID service is the same service that also handles the `Identify` protocol described above.
```